### PR TITLE
- Optimize distance calculation for tensors with single dense subspace.

### PR DESCRIPTION
**Pull request by baldersheim**
- Let EmptySubspace be invalid.
- Add noexcept to get_tensor(s).

@geirst and @havardpe PR

### CODE REVIEW

**Comment by havardpe** [comment]
I suggest using a TypedCells with size 0 rather than adding an invalid bool to signal that the cells are not present. Note that most tensor code will expect that the appropriate number of cells are present. We might need to have an empty subspace with a pointer to the appropriate number of zero values combined with a size of zero, making it work either way...

**Comment by baldersheim** [comment]
That is the approach used now. It has a proper backing of zero valued cells, but says its size is zero.

File: searchlib/src/vespa/searchlib/tensor/distance_calculator.h [hunk 1]
```diff
@@ -36,27 +36,44 @@ class DistanceCalculator {
...
+    template<bool has_single_subspace=false>
     double calc_raw_score(uint32_t docid) const {
-        auto vectors = _attr_tensor.get_vectors(docid);
-        double result = _dist_fun->min_rawscore();
-        for (uint32_t i = 0; i < vectors.subspaces(); ++i) {
-            double distance = _dist_fun->calc(vectors.cells(i));
-            double score = _dist_fun->to_rawscore(distance);
-            result = std::max(result, score);
+        if (has_single_subspace) {
+            double distance = _dist_fun->calc(_attr_tensor.get_vector(docid, 0));
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/distance_calculator.h:45]: Shouldn't we check for `cells.size == 0` here as well? If so, that case should be covered by a unit test.

Comment (geirst) [reply searchlib/src/vespa/searchlib/tensor/distance_calculator.h:45]: We should also use `max(_dist_fun->min_rawscore(), this_score)` to match existing behaviour.

**Review by geirst — APPROVED** [review]
👍  But I would like to see a unit test (if not already existing)

**Comment by baldersheim** [comment]
It was already tested, it failed when I started specifying the template based on tensor type.

**Pull request merged**